src/flutter/shell/platform/linux_embedded/window/native_window_drm_eglstream.cc

```diff
@@ -0,0 +1,299 @@
+  drm_device_ = open(deviceFilename, O_RDWR | O_CLOEXEC);
+  if (drm_device_ == -1) {
+    LINUXES_LOG(ERROR) << "Couldn't open " << deviceFilename;
+    return;
+  }
+
+  if (!ConfigureDisplay() || !ConfigureDisplayForEglstream()) {
+    return;
+  }
+
+  valid_ = true;
+}
+
+NativeWindowDrmEglstream::~NativeWindowDrmEglstream() {
+  if (drm_device_ == -1) {
+    return;
+  }
+
+  if (drm_crtc_) {
+    drmModeSetCrtc(drm_device_, drm_crtc_->crtc_id, drm_crtc_->buffer_id,
+                   drm_crtc_->x, drm_crtc_->y, &drm_connector_id_, 1,
+                   &drm_crtc_->mode);
+    drmModeFreeCrtc(drm_crtc_);
+  }
+
+  if (drm_property_blob_) {
+    drmModeDestroyPropertyBlob(drm_device_, drm_property_blob_);
+  }
+
+  if (drm_fb_) {
+    drmModeRmFB(drm_device_, drm_fb_);
+  }
+
+  close(drm_device_);
+}
+
+bool NativeWindowDrmEglstream::ShowCursor(double x, double y) {
+  // todo: implement here.
+  return true;
+}
+
+bool NativeWindowDrmEglstream::UpdateCursor(const std::string& cursor_name,
+                                            double x, double y) {
+  if (cursor_name.compare(cursor_name_) == 0) {
+    return true;
+  }
+  cursor_name_ = cursor_name;
+
+  if (cursor_name.compare(kCursorNameNone) == 0) {
+    return DismissCursor();
+  }
+
+  // todo: implement here.
+  return true;
+}
+
+bool NativeWindowDrmEglstream::DismissCursor() {
+  // todo: implement here.
+  return true;
+}
+
+std::unique_ptr<SurfaceGlDrm<uint32_t, ContextEglDrmEglstream>>
+NativeWindowDrmEglstream::CreateRenderSurface() {
+  return std::make_unique<SurfaceGlDrm<uint32_t, ContextEglDrmEglstream>>(
+      std::make_unique<ContextEglDrmEglstream>(
+          std::make_unique<EnvironmentEglDrmEglstream>()));
+}
+
+bool NativeWindowDrmEglstream::ConfigureDisplayForEglstream() {
+  if (SetDrmClientCapability() != 0) {
+    LINUXES_LOG(ERROR) << "Couldn't set drm client capability";
+    return false;
+  }
+
+  auto plane_resources = drmModeGetPlaneResources(drm_device_);
+  if (!plane_resources) {
+    LINUXES_LOG(ERROR) << "Couldn't get plane resources";
+    return false;
+  }
+  drm_plane_id_ = FindPlane(plane_resources);
+  drmModeFreePlaneResources(plane_resources);
+  if (drm_plane_id_ == -1) {
+    LINUXES_LOG(ERROR) << "Could not find a plane.";
+    return false;
+  }
+
+  auto atomic = drmModeAtomicAlloc();
+  if (!atomic) {
+    LINUXES_LOG(ERROR) << "Couldn't allocate atomic";
+    return false;
+  }
+  int ret = -1;
+  if (AssignAtomicRequest(atomic)) {
+    ret = drmModeAtomicCommit(drm_device_, atomic,
+                              DRM_MODE_ATOMIC_ALLOW_MODESET, nullptr);
+  }
+  drmModeAtomicFree(atomic);
+  if (ret != 0) {
+    LINUXES_LOG(ERROR) << "Failed to commit an atomic property change request";
+    return false;
+  }
+
+  return true;
+}
+
+int NativeWindowDrmEglstream::SetDrmClientCapability() {
+  if (drmSetClientCap(drm_device_, DRM_CLIENT_CAP_UNIVERSAL_PLANES, 1) == 0) {
```

Comment:
Please add an error message if it fails.

Reply:
I added an error message.